### ISSUE TITLE
Non-Veteran Claimants | DB migration to support unrecognized claimants and POAs

### ISSUE THREAD

**Issue by leikkisa** [department-of-veterans-affairs/caseflow#15573]
## Description
This ticket is an enabler to set up the ability to handle unrecognized appellants and power of attorneys, including the database migration and setting up the models.  This is intended to be a temporary store to unblock intake of appeals with unrecognized appellants until integration with MPI and business processes with VBA define how to process these in the long term.

Here are some examples of options:

A. Use a combination of single-table and multi-table inheritance (3 tables)
  - Upsides: Data is more clearly delineated
  - Downsides - this is a complex pattern to implement

B. Use one table, capture utilize relationship to differentiate Power of Attorney
  - Upsides: one table!
  - Downsides: We would need to do some modeling to be able to return the appellant or poa for the appellant given the decision review

C. Use polymorphism, where a parent UnrecognizedEntity table can have one UnrecognizedAppellant and one UnrecognizedPowerOfAttorney both connected to it.  This would involve three tables, but the architecture may be simpler.  Polymorphism can also be used to create two tables (one for appellants and one for POAs).  Instead of abstracting a parent class, some code would be duplicated for each.

**Option A: Use a combination of single-table and multi-table inheritance**
Appellant and POA would use multi-table inheritance from UnrecognizedEntity.  So UnrecognizedEntity would have a table, and would have two subclasses, each with their own DB table.  In turn, each of these tables would have a "type" to differentiate between Person and Organization

1. UnrecognizedEntity
    1. UnrecognizedAppellant (has "type" to differentiate organization and person)
        1. UnrecognizedPersonAppellant
        1. UnrecognizedOrganizationAppellant
    1. UnrecognizedPowerOfAttorney (has "type" to differentiate organization and person)
        1. UnrecognizedPersonPowerOfAttorney
        1. UnrecognizedOrganizationPowerOfAttorney

## Acceptance criteria
- [ ] DB migration and model container for UnrecognizedEntity
  - [ ] decision_review_type
  - [ ] decision_review_id
  - [ ] type (similar to entity_type)
  - [ ] entity_type (this would be either UnrecognizedAppellant or UnrecognizedPowerOfAttorney)
  - [ ] entity_id
- [ ] DB migration to add UnrecognizedAppellant
  - [ ] Timestamps
  - [ ] Decision review type
  - [ ] Decision review ID
  - [ ] Relationship - enumerated list to include: Attorney, Child, Parent, Spouse, Widow, Other
  - [ ] Type - UnrecognizedPersonAppellant or UnrecognizedOrganizationAppellant
  - [ ] Organization's name (nullable if type is UnrecognizedPersonAppellant)
  - [ ] First name (nullable if type is UnrecognizedPersonAppellant)
  - [ ] Middle name/initial (nullable)
  - [ ] Last name (nullable)
  - [ ] Suffix (nullable)
  - [ ] Street address 1
  - [ ] Street address 2 (nullable)
  - [ ] City
  - [ ] State (dropdown)
  - [ ] Zip
  - [ ] Country
  - [ ] Phone number (VARCHAR, nullable)
- [ ] DB migration to add UnrecognizedPowerOfAttorney following the pattern for UnrecognizedAppellant
- [ ] Create models for UnrecognizedAppellant and UnrecognizedPowerOfAttorney
  - [ ] Add subclasses for UnrecognizedPersonAppellant/PowerOfAttorney & UnrecognizedOrganization/PowerOfAttorney
- [ ] Add all Rails associations (more details TBD after discussion)

**Option B: Use one table, capture utilize relationship to differentiate Power of Attorney**

This would involve creating one table "UnrecognizedEntity", and utilize relationship to differentiate a power of attorney from an appellant.  Compared to option 1, this would not have a separate table for UnrecognizedAppellant and UnrecognizedPowerOfAttorney.  It would still have a "type" column to differentiate between person and organization.

**Option C:  Use polymorphism**
In this case, we would create UnrecognizedEntity as follows:

- [ ] DB migration for UnrecognizedEntity
  - [ ] decision_review_type
  - [ ] decision_review_id
  - [ ] unrecognized_appellant_id
  - [ ] unrecognized_power_of_attorney_id (nullable)
  - [ ] bgs_power_of_attorney_id (nullable)

UnrecognizedAppellant and UnrecognizedPowerOfAttorney models and tables would be created as described in Option A.

## Out of scope
- Handling logic to use an unrecognized entity if a claimant or power of attorney is not available

## Open questions
- Should we use the term "claimant" instead of "appellant" in order to be more generic in case this functionality gets extended to claim reviews for VBA or other lines of business?
- This proposal for an unrecognized appellant is to not add a "claimant" to the decision review and just add the unrecognized appellant instead. This would require some modeling to fall back on the unrecognized entity for appellants or POAs when otherwise not available. Another option would be to create the appropriate claimant type, and connect it to the unrecognized appellant. One challenge with this is that dependents are currently subclasses of BgsRelatedClaimant and other claimant-specific logic such as validation, so refactoring would be required.  We could add a claimant in a downstream process later, or if we do create one at the same time we could update it.
- Does UnrecognizedAppellant need to be named UnrecognizedAppellantEntity? We may need to use some Rails syntax to simplify naming.

## Technical notes
All fields added can be strings, except for the timestamps (datetime) and the decision review ID (integer).

Note: There are existing Caseflow tables for People and Organizations, but they are not appropriate for our purposes.

Unfortunately, it does not look like there's a way to set up 
[When To Use Single Table Inheritance vs Multiple Table Inheritance](https://medium.com/@User3141592/when-to-use-single-table-inheritance-vs-multiple-table-inheritance-db7e9733ae2e)
